packages/firestore/src/local/shared_client_state.ts

```diff
@@ -0,0 +1,422 @@
+ * and updates its local state as new state is observed.
+ *
+ * `SharedClientState` is primarily used for synchronization in Multi-Tab
+ * environments. Each tab is responsible for registering its active query
+ * targets and mutations. As state changes happen in other clients, the
+ * `SharedClientState` class will call back into SyncEngine to keep the
+ * local state up to date.
+ *
+ * TODO(multitab): Add callbacks to SyncEngine
+ */
+export interface SharedClientState {
+  /** Associates a new Mutation Batch ID with the current Firestore client. */
+  addLocalPendingMutation(batchId: BatchId): void;
+
+  /** Removes a Mutation Batch ID for the current Firestore client. */
+  removeLocalPendingMutation(batchId: BatchId): void;
+
+  /**
+   * Gets the minimum mutation batch for all active clients.
+   *
+   * The implementation for this may require O(n) runtime, where 'n' is the
+   * number of clients.
+   */
+  getMinimumGlobalPendingMutation(): BatchId | null;
+
+  /** Associates a new Query Target ID with the current Firestore clients. */
+  addLocalQueryTarget(targetId: TargetId): void;
+
+  /** Removes a Query Target ID for the current Firestore clients. */
+  removeLocalQueryTarget(targetId: TargetId): void;
+
+  /**
+   * Gets the active Query Targets IDs for all active clients.
+   *
+   * The implementation for this may require O(n) runtime, where 'n' is the size
+   * of the result set.
+   */
+  getAllActiveQueryTargets(): SortedSet<TargetId>;
+
+  /**
+   * Starts the SharedClientState, reads existing client data for all
+   * `knownClients` and registers listeners for updates to new and existing
+   * clients.
+   */
+  start(knownClients: ClientKey[]): void;
+
+  /** Shuts down the `SharedClientState` and its listeners. */
+  shutdown(): void;
+}
+
+/**
+ * The JSON representation of a clients's metadata as used during LocalStorage
+ * serialization. The ClientKey is omitted here as it is encoded as part of the
+ * key.
+ */
+interface ClientStateSchema {
+  lastUpdateTime: number;
+  activeTargetIds: number[];
+  minMutationBatchId: number | null;
+  maxMutationBatchId: number | null;
+}
+
+/**
+ * Metadata state of a single client. Includes query targets, the minimum
+ * pending and maximum pending mutation batch ID, as well as the last update
+ * time of this state.
+ */
+// Visible for testing.
+export interface ClientState {
+  readonly activeTargetIds: SortedSet<TargetId>;
+  readonly lastUpdateTime: Date;
+  readonly maxMutationBatchId: BatchId | null;
+  readonly minMutationBatchId: BatchId | null;
+}
+
+/**
+ * This class represents the immutable ClientState for a client read from
+ * LocalStorage. It contains the list of its active query targets and the range
+ * of its pending mutation batch IDs.
+ */
+class RemoteClientState implements ClientState {
+  constructor(
+    readonly clientKey: ClientKey,
+    public lastUpdateTime: Date,
+    public activeTargetIds: SortedSet<TargetId>,
```

Comment:
lastUpdateTime and activeTargetIds can be readonly too, I think?

Reply:
Yes, fixed.